src/custom/state/appData/atoms.ts

```diff
@@ -0,0 +1,96 @@
+import { atom } from 'jotai'
+import { atomWithStorage } from 'jotai/utils'
+
+import {
+  AddAppDataToUploadQueueParams,
+  AppDataPendingToUpload,
+  AppDataInfo,
+  FlattenedAppDataFromUploadQueue,
+  RemoveAppDataFromUploadQueueParams,
+  UpdateAppDataOnUploadQueueParams,
+} from 'state/appData/types'
+import { buildAppDataRecordKey, parseAppDataRecordKey } from 'state/appData/utils'
+
+const INITIAL_APP_DATA_UPLOAD_QUEUE: AppDataPendingToUpload = {}
+
+/**
+ * Base atom that store the current appDataInfo
+ */
+export const appDataInfoAtom = atom<AppDataInfo | null>(null)
+
+/**
+ * Base atom that stores all appData pending to be uploaded
+ */
+export const appDataUploadQueueAtom = atomWithStorage(
+  'appDataUploadQueue', // local storage key
+  INITIAL_APP_DATA_UPLOAD_QUEUE
+)
+
+/**
+ * Write only atom to add an appData to upload queue
+ */
+export const addAppDataToUploadQueueAtom = atom(
+  null,
+  (get, set, { chainId, orderId, appData }: AddAppDataToUploadQueueParams) => {
+    set(appDataUploadQueueAtom, () => {
+      const docs = get(appDataUploadQueueAtom)
+      const key = buildAppDataRecordKey({ chainId, orderId })
+
+      return {
+        ...docs,
+        [key]: { ...appData, uploading: false },
+      }
+    })
+  }
+)
+
+/**
+ * Write only atom to update upload status of an appData on upload queue
+ */
+export const updateAppDataOnUploadQueueAtom = atom(
+  null,
+  (get, set, { chainId, orderId, uploading, tryAfter }: UpdateAppDataOnUploadQueueParams) => {
+    set(appDataUploadQueueAtom, () => {
+      const docs = get(appDataUploadQueueAtom)
+      const key = buildAppDataRecordKey({ chainId, orderId })
+
+      if (!docs[key]) {
+        return docs
```

Comment:
not too familiar with jotai, but shouldn't it fail in this case? or should we set it anyway?

Reply:
What would happen on redux if you try to update a state that doesn't exist?

We don't want to throw on state update, so I just ignore and keep the same state without changes.